exotica_core/src/visualization_meshcat.cpp

```diff
@@ -0,0 +1,342 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of  nor the names of its contributors may be used to
+//    endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <exotica_core/visualization_meshcat.h>
+#include <exotica_core/visualization_meshcat_types.h>
+
+#include <chrono>
+
+#define TIMEOUT 10000
+
+namespace exotica
+{
+std::vector<double> FrameToVector(const KDL::Frame& frame, double scale_x = 1.0, double scale_y = 1.0, double scale_z = 1.0)
+{
+    std::vector<double> ret(16);
+    ret[0] = frame.M.data[0] * scale_x;
+    ret[1] = frame.M.data[3];
+    ret[2] = frame.M.data[6];
+    ret[3] = 0.0;
+    ret[4] = frame.M.data[1];
+    ret[5] = frame.M.data[4] * scale_y;
+    ret[6] = frame.M.data[7];
+    ret[7] = 0.0;
+    ret[8] = frame.M.data[2];
+    ret[9] = frame.M.data[5];
+    ret[10] = frame.M.data[8] * scale_z;
+    ret[11] = 0.0;
+    ret[12] = frame.p.data[0];
+    ret[13] = frame.p.data[1];
+    ret[14] = frame.p.data[2];
+    ret[15] = 1.0;
+    return ret;
+};
+
+std::vector<double> PositionToVector(const KDL::Frame& frame)
+{
+    std::vector<double> ret(3);
+    ret[0] = frame.p.data[0];
+    ret[1] = frame.p.data[1];
+    ret[2] = frame.p.data[2];
+    return ret;
+};
+
+std::vector<double> QuaternionToVector(const KDL::Frame& frame)
+{
+    std::vector<double> ret(4);
+    frame.M.GetQuaternion(ret[0], ret[1], ret[2], ret[3]);
+    return ret;
+};
+
+VisualizationMeshcat::VisualizationMeshcat(ScenePtr scene, const std::string& url, bool use_mesh_materials) : scene_(scene), context_(1), zmq_url_(url)
+{
+    HIGHLIGHT_NAMED("VisualizationMeshcat", "Initialising visualizer");
+    Initialize(use_mesh_materials);
+}
+VisualizationMeshcat::~VisualizationMeshcat() = default;
+
+void VisualizationMeshcat::Initialize(bool use_mesh_materials)
+{
+    ConnectZMQ();
+    web_url_ = RequestWebURL();
+    if (web_url_.size() > 7) file_url_ = web_url_.substr(0, web_url_.size() - 7) + "files/";
```

Comment:
Comment why greater 7?

Reply:
Size of the string "files/". 
Don't dwell on it though. It will be replaced with a `RequestFileURL()` method handling this on the server side when I get to push that through.